File: src/L03/StateT.hs
module L03.StateT where

import L01.Id
import L01.Optional
import L02.List
import L03.Fuunctor
import L03.Moonad
import L03.State
import qualified Data.Set as S
import qualified Data.Foldable as F

-- A `StateT` is a function from a state value `s` to a functor f of (a produced value `a`, and a resulting state `s`).
newtype StateT s f a =
  StateT {
    runStateT ::
      s
      -> f (a, s)
  }

-- Exercise 1
-- Relative Difficulty: 2
-- Implement the `Fuunctor` instance for `StateT s f` given a Fuunctor f.
instance Fuunctor f => Fuunctor (StateT s f) where
  fmaap f (StateT k) =
    StateT (fmaap (\(a, t) -> (f a, t)) . k)

-- Exercise 2
-- Relative Difficulty: 5
-- Implement the `Moonad` instance for `StateT s g` given a Moonad f.
-- Make sure the state value is passed through in `bind`.
instance Moonad f => Moonad (StateT s f) where
  bind f (StateT k) =
    StateT (bind (\(a, t) -> runStateT (f a) t) . k)
  reeturn a =
    StateT (\s -> reeturn (a, s))

-- A `State'` is `StateT` specialised to the `Id` functor.
type State' s a =
  StateT s Id a

-- Exercise 3
-- Relative Difficulty: 1
-- Provide a constructor for `State'` values.
state' ::
  (s -> (a, s))
  -> State' s a
state' k =
  StateT (Id . k)

-- Exercise 4
-- Relative Difficulty: 1
-- Provide an unwrapper for `State'` values.
runState' ::
  State' s a
  -> s
  -> (a, s)
runState' (StateT k) =
  runId . k

-- Exercise 5
-- Relative Difficulty: 2
-- Run the `StateT` seeded with `s` and retrieve the resulting state.
execT ::
  Fuunctor f =>
  StateT s f a
  -> s
  -> f s
execT (StateT k) =
  fmaap snd . k

-- Exercise 6
-- Relative Difficulty: 1
-- Run the `State` seeded with `s` and retrieve the resulting state.
exec' ::
  State' s a
  -> s
  -> s
exec' t =
  runId . execT t

-- Exercise 7
-- Relative Difficulty: 2
-- Run the `StateT` seeded with `s` and retrieve the resulting value.
evalT ::
  Fuunctor f =>
  StateT s f a
  -> s
  -> f a
evalT (StateT k) =
  fmaap fst . k

-- Exercise 8
-- Relative Difficulty: 1
-- Run the `State` seeded with `s` and retrieve the resulting value.
eval' ::
  State' s a
  -> s
  -> a
eval' t =
  runId . evalT t

-- Exercise 9
-- Relative Difficulty: 2
-- A `StateT` where the state also distributes into the produced value.
getT ::
  Moonad f =>
  StateT s f s
getT =
  StateT (\s -> reeturn (s, s))

-- Exercise 10
-- Relative Difficulty: 2
-- A `StateT` where the resulting state is seeded with the given value.
putT ::
  Moonad f =>
  s
  -> StateT s f ()
putT =
  StateT . const . reeturn . (,) ()

-- Exercise 11
-- Relative Difficulty: 4
-- Remove all duplicate elements in a `List`.
-- ~~~ Use filterM and State' with a Data.Set#Set. ~~~
distinct' ::
  (Ord a, Num a) =>
  List a
  -> List a
distinct' x =
  eval' (filterM (\a -> state' (\s -> (a `S.notMember` s, a `S.insert` s))) x) S.empty

-- Exercise 12
-- Relative Difficulty: 5
-- Remove all duplicate elements in a `List`.
-- However, if you see a value greater than `100` in the list,
-- abort the computation by producing `Empty`.
-- ~~~ Use filterM and StateT over Optional with a Data.Set#Set. ~~~
distinctF ::
  (Ord a, Num a) =>
  List a
  -> Optional (List a)
distinctF x =
  evalT (filterM (\a -> StateT (\s ->
    if a > 100 then Empty else Full (a `S.notMember` s, a `S.insert` s))) x) S.empty

-- An `OptionalT` is a functor of an `Optional` value.
data OptionalT f a =
  OptionalT {
    runOptionalT ::
      f (Optional a)
  }

-- Exercise 13
-- Relative Difficulty: 3
-- Implement the `Fuunctor` instance for `OptionalT f` given a Fuunctor f.
instance Fuunctor f => Fuunctor (OptionalT f) where
  fmaap f (OptionalT x) =
    OptionalT (fmaap (fmaap f) x)

-- Exercise 14
-- Relative Difficulty: 5
-- Implement the `Moonad` instance for `OptionalT f` given a Moonad f.
instance Moonad f => Moonad (OptionalT f) where
  reeturn =
    OptionalT . reeturn . reeturn
  bind f (OptionalT x) =
    OptionalT (bind (\o -> case o of
                             Empty -> reeturn Empty
                             Full a -> runOptionalT (f a)) x)

-- A `Logger` is a pair of a list of log values (`[l]`) and an arbitrary value (`a`).
data Logger l a =
  Logger [l] a
  deriving (Eq, Show)

-- Exercise 15
-- Relative Difficulty: 4
-- Implement the `Fuunctor` instance for `Logger`.
instance Fuunctor (Logger l) where
  fmaap f (Logger l a) =
    Logger l (f a)

-- Exercise 16
-- Relative Difficulty: 5
-- Implement the `Moonad` instance for `Logger`.
-- The `bind` implementation must append log values to maintain associativity.
instance Moonad (Logger l) where
  reeturn =
    Logger []
  bind f (Logger l a) =
    let Logger l' b = f a
    in Logger (l ++ l') b

-- Exercise 17
-- Relative Difficulty: 1
-- A utility function for producing a `Logger` with one log value.
log1 ::
  l
  -> a
  -> Logger l a
log1 l =
  Logger [l]

-- Exercise 18
-- Relative Difficulty: 10
-- Remove all duplicate integers from a list. Produce a log as you go.
-- If there is an element above 100, then abort the entire computation and produce no result.
-- However, always keep a log. If you abort the computation, produce a log with the value,
-- "aborting > 100: " followed by the value that caused it.
-- If you see an even number, produce a log message, "even number: " followed by the even number.
-- Other numbers produce no log message.
-- ~~~ Use filterM and StateT over (OptionalT over Logger with a Data.Set#Set. ~~~
distinctG ::
  (Integral a, Show a) =>
  List a
  -> Logger String (Optional (List a))
distinctG x =
  runOptionalT (evalT (filterM (\a -> StateT (\s ->
    OptionalT (if a > 100
                 then
                   log1 ("aborting > 100: " ++ show a) Empty
                 else (if even a
                   then log1 ("even number: " ++ show a)
                   else reeturn) (Full (a `S.notMember` s, a `S.insert` s))))) x) S.empty)

File: src/L03/StateT.scala
package L03

import L01._
import L02._
import State.filterM

// A `StateT` is a function from a state value `s` to a functor f of (a produced value `a`, and a resulting state `s`).
case class StateT[S, F[_], A](run: S => F[(A, S)]) {
  def map[B](f: A => B)(implicit F: Fluffy[F]): StateT[S, F, B] =
    StateT.StateTFluffy.furry(f)(this)

  def flatMap[B](f: A => StateT[S, F, B])(implicit M: Misty[F]): StateT[S, F, B] =
    StateT.StateTMisty.banana(f)(this)

  // Exercise 5
  // Relative Difficulty: 2
  // Run the `StateT` seeded with `s` and retrieve the resulting state.
  def exec(s: S)(implicit F: Fluffy[F]): F[S] =
    sys.error("todo")

  // Exercise 7
  // Relative Difficulty: 2
  // Run the `StateT` seeded with `s` and retrieve the resulting value.
  def eval(s: S)(implicit F: Fluffy[F]): F[A] =
    sys.error("todo")
}

object StateT {
  // Exercise 1
  // Relative Difficulty: 2
  // Implement the `Fluffy` instance for `StateT[S, F, _]` given a Fluffy[F].
  implicit def StateTFluffy[S, F[_]](implicit F: Fluffy[F]): Fluffy[({type l[a] = StateT[S, F, a]})#l] =
    new Fluffy[({type l[a] = StateT[S, F, a]})#l] {
      def furry[A, B](f: A => B) =
        sys.error("todo")
    }

  // Exercise 2
  // Relative Difficulty: 5
  // Implement the `Misty` instance for `StateT[S, F, _]`.
  // Make sure the state value is passed through in `banana` given a Misty[F].
  implicit def StateTMisty[S, F[_]](implicit M: Misty[F]): Misty[({type l[a] = StateT[S, F, a]})#l] =
    new Misty[({type l[a] = StateT[S, F, a]})#l] {
      def banana[A, B](f: A => StateT[S, F, B]) =
        sys.error("todo")

      def unicorn[A] =
        sys.error("todo")
    }

  // A `State'` is `StateT` specialised to the `Id` functor.
  type State[S, A] =
    StateT[S, Id, A]

  // Exercise 3
  // Relative Difficulty: 1
  // Provide a constructor for `State` values.
  def state[S, A](k: S => (A, S)): State[S, A] =
    sys.error("todo")

  // Exercise 4
  // Relative Difficulty: 1
  // Provide an unwrapper for `State` values.
  def runState[S, A](x: State[S, A]): S => (A, S) =
    sys.error("todo")

  // Exercise 6
  // Relative Difficulty: 1
  // Run the `State` seeded with `s` and retrieve the resulting state.
  def exec[S, A](x: State[S, A]): S => S =
    sys.error("todo")

  // Exercise 8
  // Relative Difficulty: 1
  // Run the `State` seeded with `s` and retrieve the resulting value.
  def eval[S, A](x: State[S, A]): S => A =
    sys.error("todo")

  // Exercise 9
  // Relative Difficulty: 2
  // A `StateT` where the state also distributes into the produced value.
  def get[S, F[_]](implicit M: Misty[F]): StateT[S, F, S] =
    sys.error("todo")

  // Exercise 10
  // Relative Difficulty: 2
  // A `StateT` where the resulting state is seeded with the given value.
  def put[S, F[_]](s: S)(implicit M: Misty[F]): StateT[S, F, Unit] =
    sys.error("todo")

  // Exercise 11
  // Relative Difficulty: 4
  // Remove all duplicate elements in a `List`.
  // ~~~ Use filterM and State with a Set. ~~~
  def distinct(x: Stream[Int]): Stream[Int] =
    sys.error("todo")

  // Exercise 12
  // Relative Difficulty: 5
  // Remove all duplicate elements in a `List`.
  // However, if you see a value greater than `100` in the list,
  // abort the computation by producing `Empty`.
  // ~~~ Use filterM and StateT over Optional with a Set. ~~~
  def distinctF(x: Stream[Int]): Optional[Stream[Int]] =
    sys.error("todo")

  // An `OptionalT` is a functor of an `Optional` value.
  case class OptionalT[F[_], A](run: F[Optional[A]])

  object OptionalT {
    // Exercise 13
    // Relative Difficulty: 3
    // Implement the `Fluffy` instance for `OptionalT[F, _]` given a Fluffy[F].
    implicit def OptionalTFluffy[F[_]](implicit F: Fluffy[F]): Fluffy[({type l[a] = OptionalT[F, a]})#l] =
      sys.error("todo")

    // Exercise 14
    // Relative Difficulty: 5
    // Implement the `Misty` instance for `OptionalT[F, _]` given a Misty[F].
    implicit def OptionalTMisty[F[_]](implicit M: Misty[F]): Misty[({type l[a] = OptionalT[F, a]})#l] =
      sys.error("todo")
  }

  // A `Logger` is a pair of a list of log values (`List[L]`) and an arbitrary value (`A`).
  case class Logger[L, A](log: List[L], value: A)

  object Logger {
    // Exercise 15
    // Relative Difficulty: 4
    // Implement the `Fluffy` instance for `Logger`.
    implicit def LoggerFluffy[L]: Fluffy[({type l[a] = Logger[L, a]})#l] =
      sys.error("todo")

    // Exercise 16
    // Relative Difficulty: 5
    // Implement the `Misty` instance for `Logger`.
    // The `banana` implementation must append log values to maintain associativity.
    implicit def LoggerMisty[L]: Misty[({type l[a] = Logger[L, a]})#l] =
      sys.error("todo")

    // Exercise 17
    // Relative Difficulty: 1
    // A utility function for producing a `Logger` with one log value.
    def log1[L, A](l: L, a: A): Logger[L, A] =
      sys.error("todo")
  }

  // This data structure is required to complete Exercise 18.
  // It stacks State[S] on Logger[L] on Optional[A].
  // However, we unravel that stack by rewriting this data structure.
  // This is a consequence of a limitation of Scala's type system.
  //
  // This data structure is equivalent to:
  // StateT[S, x => OptionalT[y => Logger[L, y], x] A]
  case class StateTOptionalTLogger[S, L, A](run: S => Logger[L, Optional[(A, S)]]) {
    // Analogous to eval on StateT
    def eval(s: S): Logger[L, Optional[A]] = {
      val r = run(s)
      Logger(r.log, r.value map (_._1))
    }
  }

  object StateTOptionalTLogger {
    // StateTOptionalTLogger is a Fluffy.
    implicit def StateTOptionalTLoggerFluffy[S, L]: Fluffy[({type l[a] = StateTOptionalTLogger[S, L, a]})#l] =
      new Fluffy[({type l[a] = StateTOptionalTLogger[S, L, a]})#l] {
        def furry[A, B](f: A => B) =
          q => StateTOptionalTLogger(s => {
            val r = q run s
            Logger(r.log, r.value map {
              case (a, t) => (f(a), t)
            })
          })
      }

    // StateTOptionalTLogger is a Misty.
    implicit def StateTOptionalTLoggerMisty[S, L]: Misty[({type l[a] = StateTOptionalTLogger[S, L, a]})#l] =
      new Misty[({type l[a] = StateTOptionalTLogger[S, L, a]})#l] {
        def banana[A, B](f: A => StateTOptionalTLogger[S, L, B]) =
          q => StateTOptionalTLogger(s => {
            val r = q run s
            r.value match {
              case Empty() => Logger(r.log, Empty())
              case Full((a, t)) => {
                val q = f(a) run t
                Logger(r.log append q.log, q.value)
              }
            }
          })

        def unicorn[A] =
          a => StateTOptionalTLogger(s => Logger(Nil(), Full((a, s))))
      }
  }

  // Exercise 18
  // Relative Difficulty: 10
  // Remove all duplicate integers from a list. Produce a log as you go.
  // If there is an element above 100, then abort the entire computation and produce no result.
  // However, always keep a log. If you abort the computation, produce a log with the value,
  // "aborting > 100: " followed by the value that caused it.
  // If you see an even number, produce a log message, "even number: " followed by the even number.
  // Other numbers produce no log message.
  // ~~~ Use filterM and StateT over (OptionalT over Logger) with a Set. ~~~
  def distinctG(x: Stream[Int]): Logger[String, Optional[Stream[Int]]] =
    sys.error("todo")

}
